shell/utils/fleet.ts

```diff
@@ -0,0 +1,146 @@
+import { BundleDeploymentResource, ModifiedStatus, NonReadyBundle, ResourceKey } from '@shell/types/resources/fleet';
+import { STATES_ENUM } from '@shell/plugins/dashboard-store/resource-class';
+import { FLEET as FLEET_ANNOTATIONS } from '@shell/config/labels-annotations';
+
+type state = STATES_ENUM.READY | STATES_ENUM.MODIFIED | STATES_ENUM.MISSING | STATES_ENUM.ORPHANED;
+
+interface Resource extends BundleDeploymentResource {
+  state: state,
+}
+
+interface BundleDeploymentStatus {
+  resources?: BundleDeploymentResource[],
+  modifiedStatus?: ModifiedStatus[],
+}
+
+interface BundleStatusSummary {
+  nonReadyResources?: NonReadyBundle[],
+}
+
+interface BundleStatus {
+  resourceKey?: ResourceKey[],
+  summary?: BundleStatusSummary,
+}
+
+function resourceKey(r: BundleDeploymentResource): string {
+  return `${ r.kind }/${ r.namespace }/${ r.name }`;
+}
+
+/**
+ * bundleDeploymentResources extracts the list of resources deployed by a BundleDeployment
+ */
+export function bundleDeploymentResources(status: BundleDeploymentStatus): Resource[] {
+  // status.resources includes of resources that were deployed by Fleet *and still exist in the cluster*
+  // Use a map to avoid `find` over and over again
+  const resources = (status?.resources || []).reduce((res, r) => {
+    res[resourceKey(r)] = Object.assign({ state: STATES_ENUM.READY }, r);
+
+    return res;
+  }, {});
+
+  const modified = [];
+
+  for (const r of status?.modifiedStatus || []) {
+    const state = r.missing ? STATES_ENUM.MISSING : r.delete ? STATES_ENUM.ORPHANED : STATES_ENUM.MODIFIED;
+    const found = resources[resourceKey(r)];
+
+    // Depending on the state, the same resource can appear in both fields
+    if (found) {
+      found.state = state;
+    } else {
+      modified.push(Object.assign({ state }, r));
+    }
+  }
+
+  return modified.concat(Object.values(resources));
+}
+
+/**
+ * bundleResources extracts the list of resources deployed by a Bundle
+ */
+export function bundleResources(status: BundleStatus): Resource[] {
+  const newCounter = (): Object<string, number> => ({
```

Comment:
Pulling this all out above is neater and avoids creating a fn everytime the outer function is called

```
interface Counts { [state: string]: number}

const newCounter = (): Counts => ({
  [STATES_ENUM.READY]:    0,
  [STATES_ENUM.MISSING]:  0,
  [STATES_ENUM.ORPHANED]: 0,
  [STATES_ENUM.MODIFIED]: 0,
});
```

this also helps with the comment below...